gprofiler/profilers/java.py

```diff
@@ -259,11 +287,77 @@ def make_application_metadata(self, process: Process) -> Dict[str, Any]:
         # that loads other libs.
         libjvm_elfid = get_mapped_dso_elf_id(process, "/libjvm")
 
-        metadata = {"java_version": version, "libjvm_elfid": libjvm_elfid}
+        metadata = {
+            "java_version": version,
+            "libjvm_elfid": libjvm_elfid,
+            "jvm_flags": self.get_jvm_flags_serialized(process),
+        }
 
         metadata.update(super().make_application_metadata(process))
         return metadata
 
+    def get_jvm_flags_serialized(self, process: Process) -> Optional[List[Dict]]:
+        jvm_flags = self.get_jvm_flags(process)
+        if jvm_flags is None:
+            return None
+
+        return [flag.to_dict() for flag in sorted(jvm_flags, key=lambda flag: flag.name)]
+
+    def get_jvm_flags(self, process: Process) -> Optional[List[JvmFlag]]:
+        jvm_raw_flags: Optional[List[JvmFlag]] = self.get_jvm_flags_raw(process)
+
+        if jvm_raw_flags is None:
+            return None
+
+        elif self.jvm_flags_to_collect == JavaFlagCollectionOptions.ALL:
+            return jvm_raw_flags
+        elif self.jvm_flags_to_collect == JavaFlagCollectionOptions.DEFAULT:
+            return [flag for flag in jvm_raw_flags if self.filter_jvm_flag(flag)]
+        elif isinstance(self.jvm_flags_to_collect, list):
+            requested_flags = [flag for flag in jvm_raw_flags if flag.name in self.jvm_flags_to_collect]
+            not_found_flags = [
+                flag_name
+                for flag_name in self.jvm_flags_to_collect
+                if flag_name not in [flag.name for flag in requested_flags]
+            ]
+
+            if not_found_flags:
+                logger.error("Flags not found in jvm flags:", not_found_flags=not_found_flags)
+
+            return requested_flags
+        else:
+            logger.error("Unrecognized jvm_flags_to_collect value:", jvm_flags_to_collect=self.jvm_flags_to_collect)
+            return None
```

Comment:
As before, if getting here is a programming error and not an actual flow, either remove (and depend on type checking) or convert to an assertion.